[user-streams] Add stream support to scheduler

mlazos opened the pull request:
Summary

  Add CUDA stream context management to the inductor scheduler's codegen phase. This is the
  foundational infrastructure for multi-stream code generation: it tracks which stream is currently
  active and generates the appropriate with torch.cuda.stream(...) enter/exit wrappers around
  kernels.

  Key pieces:
  - _current_stream_ctx field tracks the active EnterCudaStreamContextLine during codegen
  - current_stream_idx / current_stream_name properties expose the current stream state (also used by
   the wrapper's nesting guard)
  - generate_stream_ctx_switching() is the main entry point called per-node during codegen — it
  compares the node's assigned stream to the current stream and emits enter/exit code only when
  switching is needed. NopKernelSchedulerNodes inherit the previous stream context since they
  generate no kernel code.

  This commit adds the codegen-time stream management; the stream assignment logic (node_to_stream,
  _populate_stream_assignments) and the codegen callsite are in the next commit in the stack.

  Test plan

  - Tests are in the third commit of the stack (test/inductor/test_user_streams.py)


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #174224
* #174223
* __->__ #165505



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

mlazos committed:
[user-streams] Add stream support to scheduler

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

eellison reviewed (commented):
one comment on design prior to full review

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

New State Added to Scheduler                                                                
                                                            
  - event_factory: Creates CUDA event symbols for synchronization                             
  - buff_to_event: Maps buffer names → CUDA events that signal when the buffer is ready
  - unjoined_events: Tracks events on each stream that haven't been synchronized yet (need
  cleanup at program exit)
  - buffers_requiring_device_check: Buffers that may be on CPU and need runtime device checks
  before record_stream
  - _current_stream_ctx: Tracks which stream context codegen is currently inside

  Properties for Stream Context Access

  - current_stream_idx / current_stream_name: Get the active stream during scheduling
  - buffers_recorded_on_current_stream: Track which buffers have already had record_stream
  called

  Core Functions

  1. get_cross_stream_dependencies(node): Analyzes a node's reads to determine:
    - Which CUDA events from other streams must be waited on
    - Which buffers from other streams need record_stream calls
  2. generate_stream_ctx_enter/exit: Emit code to enter/exit a CUDA stream context with proper
   event waits and buffer recording
  3. propagate_cross_stream_dependencies(node): For nodes scheduled mid-context, propagates
  their cross-stream sync requirements to the context entrance
  4. generate_stream_ctx_switching(node): The main dispatch logic - decides whether to stay on
   current stream, enter a new stream, or switch streams based on node's assigned stream vs
  current stream
  5. register_downstream_event(node): After a node executes, records a CUDA event marking
  completion and updates tracking structures
  6. get_final_events_to_sync(): At program end, identifies which events on side streams need
  synchronization back to the main stream




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos commented on torch/_inductor/scheduler.py:
```diff
@@ -6989,6 +7010,310 @@ def update_zero_dim_cpu_tensor(self) -> None:
                     ):
                         V.graph.zero_dim_cpu_tensor_list.add(read.name)
```
Removed

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

 Summary                                                                                            
                                                                                                     
  Remove the scheduler's cross-stream dependency computation (events to wait on, buffers to          
  record_stream). This infrastructure was computing synchronization from data dependencies, but it's
  unnecessary — users already express synchronization explicitly via event.record() / event.wait() in
   their code, which Dynamo traces as torch.ops.streams.* nodes in the FX graph.

  The scheduler now only tracks which stream context it's currently emitting code into
  (_current_stream_ctx) and generates with torch.cuda.stream(...) blocks accordingly. Cross-stream
  synchronization is left to the user's explicit event operations.

  Removed:
  - torch/_inductor/event.py (CudaEventSym, CudaEventFactory, _CudaEventRecordLine,
  _CudaEventWaitLine)
  - ENTRANCE_EVENT and EVENT_NAME_TEMPLATE constants
  - Scheduler fields: event_factory, buff_to_event, unjoined_events
  - Scheduler methods: get_cross_stream_dependencies, propagate_cross_stream_dependencies,
  register_downstream_event, get_final_events_to_sync, clear_unjoined_events, get_last_event
  - Wrapper methods: codegen_events_wait_stream, codegen_buffers_record_stream

  Simplified:
  - codegen_cuda_stream_enter no longer takes upstream_events or buffers_from_other_streams
  - generate_stream_ctx_switching no longer propagates dependencies on same-stream nodes

  -551 lines, +3 lines.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

 Summary                                                                                            
                                                                                                     
  Remove the scheduler's cross-stream dependency computation (events to wait on, buffers to          
  record_stream). This infrastructure was computing synchronization from data dependencies, but it's
  unnecessary — users already express synchronization explicitly via event.record() / event.wait() in
   their code, which Dynamo traces as torch.ops.streams.* nodes in the FX graph.

  The scheduler now only tracks which stream context it's currently emitting code into
  (_current_stream_ctx) and generates with torch.cuda.stream(...) blocks accordingly. Cross-stream
  synchronization is left to the user's explicit event operations.

  Removed:
  - torch/_inductor/event.py (CudaEventSym, CudaEventFactory, _CudaEventRecordLine,
  _CudaEventWaitLine)
  - ENTRANCE_EVENT and EVENT_NAME_TEMPLATE constants
  - Scheduler fields: event_factory, buff_to_event, unjoined_events
  - Scheduler methods: get_cross_stream_dependencies, propagate_cross_stream_dependencies,
  register_downstream_event, get_final_events_to_sync, clear_unjoined_events, get_last_event
  - Wrapper methods: codegen_events_wait_stream, codegen_buffers_record_stream

  Simplified:
  - codegen_cuda_stream_enter no longer takes upstream_events or buffers_from_other_streams
  - generate_stream_ctx_switching no longer propagates dependencies on same-stream nodes

  -551 lines, +3 lines.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

Summary

  Add CUDA stream context management to the inductor scheduler's codegen phase. This is the
  foundational infrastructure for multi-stream code generation: it tracks which stream is currently
  active and generates the appropriate with torch.cuda.stream(...) enter/exit wrappers around
  kernels.

  Key pieces:
  - _current_stream_ctx field tracks the active EnterCudaStreamContextLine during codegen
  - current_stream_idx / current_stream_name properties expose the current stream state (also used by
   the wrapper's nesting guard)
  - generate_stream_ctx_switching() is the main entry point called per-node during codegen — it
  compares the node's assigned stream to the current stream and emits enter/exit code only when
  switching is needed. NopKernelSchedulerNodes inherit the previous stream context since they
  generate no kernel code.

  This commit adds the codegen-time stream management; the stream assignment logic (node_to_stream,
  _populate_stream_assignments) and the codegen callsite are in the next commit in the stack.

  Test plan

  - Tests are in the third commit of the stack (test/inductor/test_user_streams.py)





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

Summary

  Add CUDA stream context management to the inductor scheduler's codegen phase. This is the
  foundational infrastructure for multi-stream code generation: it tracks which stream is currently
  active and generates the appropriate with torch.cuda.stream(...) enter/exit wrappers around
  kernels.

  Key pieces:
  - _current_stream_ctx field tracks the active EnterCudaStreamContextLine during codegen
  - current_stream_idx / current_stream_name properties expose the current stream state (also used by
   the wrapper's nesting guard)
  - generate_stream_ctx_switching() is the main entry point called per-node during codegen — it
  compares the node's assigned stream to the current stream and emits enter/exit code only when
  switching is needed. NopKernelSchedulerNodes inherit the previous stream context since they
  generate no kernel code.

  This commit adds the codegen-time stream management; the stream assignment logic (node_to_stream,
  _populate_stream_assignments) and the codegen callsite are in the next commit in the stack.

  Test plan

  - Tests are in the third commit of the stack (test/inductor/test_user_streams.py)





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add stream support to scheduler"

Summary

  Add CUDA stream context management to the inductor scheduler's codegen phase. This is the
  foundational infrastructure for multi-stream code generation: it tracks which stream is currently
  active and generates the appropriate with torch.cuda.stream(...) enter/exit wrappers around
  kernels.

  Key pieces:
  - _current_stream_ctx field tracks the active EnterCudaStreamContextLine during codegen
  - current_stream_idx / current_stream_name properties expose the current stream state (also used by
   the wrapper's nesting guard)
  - generate_stream_ctx_switching() is the main entry point called per-node during codegen — it
  compares the node's assigned stream to the current stream and emits enter/exit code only when
  switching is needed. NopKernelSchedulerNodes inherit the previous stream context since they
  generate no kernel code.

  This commit adds the codegen-time stream management; the stream assignment logic (node_to_stream,
  _populate_stream_assignments) and the codegen callsite are in the next commit in the stack.

  Test plan

  - Tests are in the third commit of the stack (test/inductor/test_user_streams.py)





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]
